hw/top_earlgrey/data/pins_nexysvideo.xdc

```diff
@@ -11,6 +11,11 @@ create_generated_clock -name clk_50_unbuf -source [get_pin clkgen/pll/CLKIN1] [g
 create_generated_clock -name clk_48_unbuf -source [get_pin clkgen/pll/CLKIN1] [get_pin clkgen/pll/CLKOUT1]
 set_clock_groups -group clk_50_unbuf -group clk_48_unbuf -asynchronous
 
+## Preserve prim_prince modules and setup multi-cycle paths
+set_property KEEP_HIERARCHY TRUE [get_cells top_earlgrey/u_flash_eflash/gen_flash_banks[*].i_core/u_scramble/u_cipher]
+set_multicycle_path -setup 2 -through [get_pins -of_objects [get_cells top_earlgrey/u_flash_eflash/gen_flash_banks[*].i_core/u_scramble/u_cipher]]
```

Comment:
it's been a long time, but my recollection is that usually you would need a `set_multicycle_path -hold 1` with the same args as your existing setup MCP, unless you still want the hold check on the zero cycle (which might be the case here, I haven't thought about it much).  (Basically my MCP files always had this n/n-1 pattern to them.)

Reply:
you are right :)
I put the comment at a bad place.  If you scroll past our comments you will see the hold `hold` case.

Reply:
Hah - the way this wrapped on my screen I didn't see the next line.  I guess I was correct. :)

Reply:
> @cdgori
> @msfschaffner
> @eunchan
> 
> does this look right to you guys? the `prim_prince` module is completely combo, so i figured making it multicycle path through all pins should be okay.

Just double checking, all of the inputs are stable for the 2 clocks (key_i/data_i/dec_i)?  No chance that dec_i is late arriving or something like that?

Reply:
yeah the inputs data and dec are coming from fifo'd outputs and not popped until descrambling.
key is hardwired at the moment..but it should long term come from otp flops that are static after initial sense.

Honestly...i don't like relying on multi-cycle path to make things work... there's just no clean way to ensure design changes propagate to constraints... i'd rather have things be correct by construction.. and you can see I have a giant comment regarding that.

Reply:
> you are right :)
> I put the comment at a bad place. If you scroll past our comments you will see the hold `hold` case.

I removed my comment, since I did eventually scroll down and see you had done it already :)

Reply:
> yeah the inputs data and dec are coming from fifo'd outputs and not popped until descrambling completes.
> key is hardwired at the moment..but it should long term come from otp flops that are static after initial sense.
> 
> Honestly...i don't like relying on multi-cycle path to make things work... there's just no clean way to ensure design changes propagate to constraints... i'd rather have things be correct by construction.. and you can see I have a giant comment regarding that.

I hate them and try to remove them everywhere I can (sometimes you just can't but I try).  They can easily get you a dead chip because nothing is keeping constraints in sync with design as you say, and nothing is simulating constraints.

Reply:
@msfschaffner .. say.... should we consider pipelining inside `prim_present` and `prim_prince` directly? :)
i ended up going this approach for `gf_mult`... it is more expensive to do (since you need storage inside).. and there are limits to how you can carve it (should probably be divisible).. but you know..no need for constraint tweaking :)

I have to admit though, if not for constraints, the multi-cycle approach in this case is pretty appealing, since it would also solve another problem i had to work-around.

Reply:
actually, I've been thinking about placing a register halfway into `prim_prince` in order to optimize timing in #2655. would that work for you as well? since the key is static, we probably do not have to pipeline the key into the second cycle in `prim_prince`, hence you only need 64 flops per `prim_prince` instance.

Reply:
that would definitely work for me :) Are you thinking about placing a flop at the reflect point?

Reply:
yes exactly - see https://github.com/lowRISC/opentitan/pull/2678.